design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,778 @@
+`resources` array or earlier functions in the `functions` array. Furthermore, a
+function could also be used to implement 'side effects' such as triggering a
+replication or backup.
+
+Below is a more detailed example of an entry in the `functions` array.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v2alpha1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: XPostgreSQLInstance
+  functions:
+  - name: my-cool-function
+    type: Container
+    # Configuration specific to `type: Container` functions.
+    container:
+      # The OCI image to pull and run.
+      image: xkpg.io/my-cool-function:0.1.0
+      # Whether to pull the function Never, Always, or IfNotPresent.
+      imagePullPolicy: IfNotPresent
+      # Secrets used to pull from a private registry.
+      imagePullSecrets:
+      - namespace: crossplane-system
+        name: my-xpkg-io-creds
+      # Note that only resource limits are supported - not requests.
+      # The function will be run with the specified resource limits.
+      resources:
+        limits:
+          memory: 64Mi
+          cpu: 250m
+      # Defaults to 'Isolated' - i.e an isolated network namespace.
+      network: Accessible
+      # How long the function may run before it's killed. Defaults to 10s.
+      timeout: 30s
+      # Containers are run by an external process listening at the supplied
+      # endpoint. Specifying an endpoint is optional; the endpoint defaults to
+      # the below value.
+      runner:
+        endpoint: unix:///@crossplane/fn/default.sock
+    # An x-kubernetes-embedded-resource RawExtension (i.e. an unschemafied
+    # Kubernetes resource). Passed to the function as the config block of its
+    # FunctionIO.
+    config:
+      apiVersion: database.example.org/v1alpha1
+      kind: Config
+      metadata:
+        name: cloudsql
+      spec:
+        version: POSTGRES_9_6
+```
+
+### Function API
+
+This document proposes that each function uses a `FunctionIO` type as its input
+and output. In the case of `Container` functions this would correspond to stdin
+and stdout. Crossplane would be responsible for reading stdout from the final
+function and applying its changes to the relevant XR and composed resources.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: FunctionIO
+config:
+  apiVersion: database.example.org/v1alpha1
+  kind: Config
+  metadata:
+    name: cloudsql
+  spec:
+    version: POSTGRES_9_6
+observed:
+  composite:
+    resource:
+      apiVersion: database.example.org/v1alpha1
+      kind: XPostgreSQLInstance
+      metadata:
+        name: my-db
+      spec:
+        parameters:
+          storageGB: 20
+        compositionSelector:
+          matchLabels:
+            provider: gcp
+      status:
+        conditions:
+        - type: Ready
+          status: True
+    connectionDetails:
+    - name: uri
+      value: postgresql://db.example.org:5432
```

Comment:
@turkenh We can currently get, but not list connection details, right? Is it feasible to extend the connection details API with list support? I believe we'd need to do so to build this.

Reply:
Do you mean listing secrets or connection details (i.e. k/v pairs in a given secret) ?

We get a [secret](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/connection/store/store.go#L43) with [FetchConnection](https://github.com/crossplane/crossplane-runtime/blob/bce61005a175d286ef5a7a75c4f83fcd273ec6dd/pkg/connection/manager.go#L128) which contains all connection details of an MR/XR as key values. 

This should be enough here I guess.